### DISCUSSION TITLE
[실험]CHAN-DST 논문 정리

### DISCUSSION THREAD

**Discussion by dkswndms4782** [bcaitech1/p3-dst-chatting-day#47] [Experiments]
## CHAN-DST

- slot imbalance 문제를 해결하고자 adaptive objective를 도입.
- a contextual hierarchical attention network (CHAN)를 사용
: dislogue history에서 relevant context를 찾기 위함. 
→ 각 턴의 발화로부터 word-level 관련 정보 검색
→ contextual representation으로 encode
→ 모든 context표현을 turn-level관련 정보로 집계한 후 word-level 정보와 합친 output 생성.
- state transition prediction task

---

## Definition
- <img src="https://render.githubusercontent.com/render/math?math=T"> : turn
- <img src="https://render.githubusercontent.com/render/math?math=U_t"> : user utterance of turn t
- <img src="https://render.githubusercontent.com/render/math?math=R_t"> : system response of turn t
- <img src="https://render.githubusercontent.com/render/math?math=X"> : {<img src="https://render.githubusercontent.com/render/math?math=(U_1, R_1), ... , (U_T, R_T)"> }
- <img src="https://render.githubusercontent.com/render/math?math=B_t"> : {<img src="https://render.githubusercontent.com/render/math?math=(s, v_t), s \in S">}
- <img src="https://render.githubusercontent.com/render/math?math=S"> : set of slots
<img src="https://render.githubusercontent.com/render/math?math=v_t"> : corresponding value of the slot <img src="https://render.githubusercontent.com/render/math?math=s">
slot : concatenation of a domain name and a slot name

---

## Contextual Hierarchical Attention Network
![Untitled (1)](https://user-images.githubusercontent.com/59716219/117595672-6b592f00-b17c-11eb-8e10-712d63db4c0d.png)
### 1. Sentence Encoder
![Untitled (2)](https://user-images.githubusercontent.com/59716219/117595679-701de300-b17c-11eb-86da-1d9df04a53e2.png)

`utterance encoder`
- BERT special token사용
→ [CLS] : 문장의 representation들을 합치기위해 사용 (to aggregate the whole representation of a sentence)
→ [SEP] : 문장의 끝을 나타내기위해 사용.
- <img src="https://render.githubusercontent.com/render/math?math=U_t"> = {<img src="https://render.githubusercontent.com/render/math?math=w_1^u, ..., w_l^u">} (user utterance)
<img src="https://render.githubusercontent.com/render/math?math=R_t"> = {<img src="https://render.githubusercontent.com/render/math?math=w_1^r, ..., w_{l'}^r">} (system response)
t : dialogue turn
- <img src="https://render.githubusercontent.com/render/math?math=h_t"> = <img src="https://render.githubusercontent.com/render/math?math=BERT_{finetune}([R_t;U_t])">
(<img src="https://render.githubusercontent.com/render/math?math=h_t"> : contextual word representations)
- 여기서 BERT finetune은 training도중 finetuning이 될것을 의미.

`slot-value encoder`
- <img src="https://render.githubusercontent.com/render/math?math=BERT_{fixed}">는 contextual semantics vectors로 encode해준다.
- utterance encode할때와 다른 점은 [CLS] 토큰의 output vector를 전체 문장 representation할때 사용한다. (to obtain the whole sentence representation)
- <img src="https://render.githubusercontent.com/render/math?math=h_s"> = <img src="https://render.githubusercontent.com/render/math?math=BERT_{fixed}(s)">
<img src="https://render.githubusercontent.com/render/math?math=h_t^v"> = <img src="https://render.githubusercontent.com/render/math?math=BERT_{fixed}(v_t)">
- <img src="https://render.githubusercontent.com/render/math?math=BERT_{fixed}">는 training 도중 고정되어있다. 그래서 우리 모델은 unseen slots and values에 대해서 original BERT representation로 확장해서 보는게 가능하다.

### 2. Slot-Word Attention
![Untitled (3)](https://user-images.githubusercontent.com/59716219/117595717-81ff8600-b17c-11eb-80e1-1076c957d14b.png)

- slot-word attention은 multi-head attention을 사용한다.
- <img src="https://render.githubusercontent.com/render/math?math=c_{s,t}^{word}"> = MultiHead(<img src="https://render.githubusercontent.com/render/math?math=h^s, h_t, h_t">)

### 3. Context Encoder
![Untitled (4)](https://user-images.githubusercontent.com/59716219/117595751-9ba0cd80-b17c-11eb-908a-376d3266884c.png)

- context encoder : unidirectional transformer encoder
- {1, ..., t} 턴에서 추출 된 word-level slot-related 정보의 contextual relevance를 모델링하기 위한 것.
- <img src="https://render.githubusercontent.com/render/math?math=N">개의 idenctical한 layer가 있다.
    - 각 layer는 2개의 sub-layer를 가지고 있다.
    - 첫번째 sub-layer : masked multi-head self-attention(Q = K = V)
    - 두번째 sub-layer : position-wise fully connected feed-forward network(FFN) (two linear transformations, RELU activation으로 구성)
    - <img src="https://render.githubusercontent.com/render/math?math=FFN(x) = max(0, x W_1 + b_1) W_2 + b_2">
- <img src="https://render.githubusercontent.com/render/math?math=m^n = FFN(MultiHead(m^{n-1}, m^{n-1}, m^{n-1}))">
- <img src="https://render.githubusercontent.com/render/math?math=m^0 = [c_{s,1}^{word} + PE(1), ..., c_{s,t}^{word} + PE(t)]">
- <img src="https://render.githubusercontent.com/render/math?math=c_{s,\leq t}^{ctx} = m^N">
- <img src="https://render.githubusercontent.com/render/math?math=m^n"> : n번째 context encoder레이어의 아웃풋
- <img src="https://render.githubusercontent.com/render/math?math=PE(.)"> : positional encoding function

### 4. Slot-Turn Attention
![Untitled (5)](https://user-images.githubusercontent.com/59716219/117595792-b2dfbb00-b17c-11eb-8c50-b1378364812c.png)

- turn-level relevant information을 contextual representation에서 검출해내기 위해 사용
- <img src="https://render.githubusercontent.com/render/math?math=c_{s,t}^{turn} = MultiHead(h^s, c_{s,\leq t}^{ctx},c_{s,\leq t}^{ctx})">
- 이로인해 word-level and turn-level 의 relevant information을 historical dialogues에서 얻어낼 수 있다.

### 5. Global-Local Fusion Gate
![Untitled (6)](https://user-images.githubusercontent.com/59716219/117595806-bb37f600-b17c-11eb-80db-816fd34ec100.png)

- global context와 local utterance의 균형을 맞추기 위해, contextual information과 current turn information의 비율을 조절함. 
- <img src="https://render.githubusercontent.com/render/math?math=c_{s,t}^{word},~~ c_{s,t}^{turn}">에 따라 global과 local정보가 어떻게 결합되어야할지 알려주는 fusion gate mechanism을 사용
- <img src="https://render.githubusercontent.com/render/math?math=g_{s,t} = \sigma(W_g \bigodot \[c_{s,t}^{word};c_{s,t}^{turn} ] )">
- <img src="https://render.githubusercontent.com/render/math?math=c_{s,t}^gate = g_{s,t}\bigotimes c_{s,t}^{word} + (1-g_{s,t}\bigotimes c_{s,t}^{turn})">
    - <img src="https://render.githubusercontent.com/render/math?math=W_g \in R^{2d\times d}">
    - <img src="https://render.githubusercontent.com/render/math?math=\sigma"> : Sigmoid
    - <img src="https://render.githubusercontent.com/render/math?math=\bigodot, \bigotimes"> :
- <img src="https://render.githubusercontent.com/render/math?math=o_{s,t}"> = LayerNorm(Linear(Dropout(<img src="https://render.githubusercontent.com/render/math?math=c_{s,t}^{gate}">)))
- **value <img src="https://render.githubusercontent.com/render/math?math=v_t">에 대한 확률분포와 training objective**
<img src="https://render.githubusercontent.com/render/math?math=p(v_t|U_{\leq t},~R_{\leq t}, s)"> = <img src="https://render.githubusercontent.com/render/math?math=exp(-||o_{s,t} - h_t^v||_2)\over {\sum_{v'\in V_s}exp(-||o_{s,t}-h_t^{v'}||_2)}">
<img src="https://render.githubusercontent.com/render/math?math=L_{dst}"> = <img src="https://render.githubusercontent.com/render/math?math=\sum_{s\in S}\sum ^T_{t = 1}-log(p(\hat v_t|U_{\leq t},~R_{\leq t}, s))">
    - <img src="https://render.githubusercontent.com/render/math?math=V_s"> : candidate value set of slot s
    - <img src="https://render.githubusercontent.com/render/math?math=\hat v_t \in V_s"> : ground-truth value of slot s

---

## State Transition Prediction
![Untitled (7)](https://user-images.githubusercontent.com/59716219/117595812-bffcaa00-b17c-11eb-8a74-a04347f91940.png)

- relevant context를 더 잘 포착하기 위해, auxiliary binary classification task사용. 
- <img src="https://render.githubusercontent.com/render/math?math=c_{s,t}^{stp}"> = tanh(<img src="https://render.githubusercontent.com/render/math?math=W_c \odot c_{s,t}^{gate}">)
- <img src="https://render.githubusercontent.com/render/math?math=p_{s,t}^{stp}"> = <img src="https://render.githubusercontent.com/render/math?math=\sigma (W_p \odot [c_{s,t}^{stp};c_{s, t-1}^{stp}])">
    - <img src="https://render.githubusercontent.com/render/math?math=W_c \in \R^{d\times d}">
    - <img src="https://render.githubusercontent.com/render/math?math=W_c \in \R^{2d}">
    - t = 1일때는 <img src="https://render.githubusercontent.com/render/math?math=c_{s,t}^{stp}">와 zero vectors를 concat함.
- binary CE loss (<img src="https://render.githubusercontent.com/render/math?math=y_{s,t}^{stp}"> : ground-truth transition labels // <img src="https://render.githubusercontent.com/render/math?math=p_{s,t}^{stp}"> : transition probability)
- <img src="https://render.githubusercontent.com/render/math?math=L_{stp}"> = <img src="https://render.githubusercontent.com/render/math?math=\sum_{s\in S}\sum_{t = 1}^T -y_{s,t}^{stp}~.~log(p_{s,t}^{stp})">

---

## Adaptive Objective

- hard slots와 samples에 관한 optimization을 encourage한다.
- all slots의 learning을 balancing함.
- <img src="https://render.githubusercontent.com/render/math?math=acc_s^{val}"> : accuracy of slot s on validation set
- `slot-level difficulty`
if <img src="https://render.githubusercontent.com/render/math?math=acc_s^{val} \leq acc_{s'}^{val}"> ; 
→ slot s 가 slot s'보다 더 어려운 것. 
→ <img src="https://render.githubusercontent.com/render/math?math=\alpha">  : slot-level difficulty
    - <img src="https://render.githubusercontent.com/render/math?math=\alpha_s"> = <img src="https://render.githubusercontent.com/render/math?math={1 - acc_s^{val} \over {\sum_{s'\in S} 1-acc_{s'}^{val}}} . |S|">
- `sample-level difficulty`
→ Suppose there are two samples {<img src="https://render.githubusercontent.com/render/math?math=(U_t, R_t),(s, v_t)">} and {<img src="https://render.githubusercontent.com/render/math?math=(U_{t'}, R_{t'}),(s', v_{t'})">}.
→ 만약 former confidence 가 latter보다 더 낮다면, 첫번째 sample이 두번째보다 더 어려운 것. 
→ <img src="https://render.githubusercontent.com/render/math?math=\beta"> : sample level difficulty
    - <img src="https://render.githubusercontent.com/render/math?math=\beta(s, v_t) = (1 - p(s, v_t))^\gamma">
    - <img src="https://render.githubusercontent.com/render/math?math=p(s,v_t)"> : confidence of sample <img src="https://render.githubusercontent.com/render/math?math=(U_t, R_t),(s, v_t)">
    <img src="https://render.githubusercontent.com/render/math?math=\gamma"> : hyper-parameter
- <img src="https://render.githubusercontent.com/render/math?math=L_{adapt}(s,v_t) = -\alpha_s\beta(s, v_t)log(p(s, v_t))">
- slot s가 평균 slot의 difficulty보다 높다면, <img src="https://render.githubusercontent.com/render/math?math=\alpha_s">는 s에 대한 loss를 키울 것이다. 비슷하게, sample의 optimization이 low confidence를 갖고 있다면 loss는 커질것이다.

---

## Optimization

- During joint training, we optimize the sum of these two loss functions as following
<img src="https://render.githubusercontent.com/render/math?math=L_{joint} = L_{dst} + L_{stp}">
- At the fine-tuning phase, we adopt the adaptive objective to fine-tune DST task as following:
<img src="https://render.githubusercontent.com/render/math?math=L_{finetune} = \sum_{s\in S}\sum^T_{t=1}L_{adapt(s, \hat v_t)}">
